image-hd.c

```diff
@@ -720,16 +720,14 @@ static int hdimage_setup(struct image *image, cfg_t *cfg)
 					part->name);
 			return -EINVAL;
 		}
-		if (part->offset && part->in_partition_table) {
-			if (hd->gpt) {
-				int ret = check_overlap(image, part);
-				if (ret)
-					return ret;
-			} else if (now > part->offset) {
-				image_error(image, "part %s overlaps with previous partition\n",
-					    part->name);
-				return -EINVAL;
-			}
+		if (hd->gpt) {
+			int ret = check_overlap(image, part);
+			if (ret)
+				return ret;
+		} else if (now > part->offset) {
+			image_error(image, "part %s overlaps with previous partition\n",
+				    part->name);
+			return -EINVAL;
```

Comment:
But the bootloader must be allowed to overlap with the MBR. The bootloader image must start at the beginning of the device. But is has a "don't care" area where the actual partition table is located. So we write the bootloader image first and later overwrite part of it with the MBR.
And since the MBR is now a fake partition, the overlap check will fail in this case.

Reply:
Right, see above, I think that should be solved by treating the two parts of the first 512 bytes as the two distinct things that it really is: some executable code, and a partition table. But perhaps one has "pre-cooked" binary blobs of 512 bytes that are supposed to be used as-is (knowing that a small part of the end will be overwritten)? That would ruin the fun, I guess. I must admit I've never worked with images/devices where the bootloader was actually placed in the first part of the disk.

There were no tests failing after I introduced that overlap test, so I'd need to see some examples of how the bootloader binary is meant to be included in the final image. Just thinking out loud, if one has a 512 byte bootloader image, perhaps we could create a new image-type "subimage" that would simply allow one to extract a range of bytes from a given file, so one could do
```
partition bootloader { offset = 0; image = "bootloader"; in-partition-table: "false"; }
image bootloader { subimage { image = "the-binary-blob"; offset = 0; length = 440; } }
```

Reply:
So the bootloader images that I know about expect to be written at offset zero and are usually a few hundred kilobyte large.
The area for the partition table in the first sector and the second sector gpt header are unused but still part of the image.
That's the only overlap that I know of, but maybe this is used for other stuff as well. So how about something like this:

`allow-overlap = true/false/header`

`true` means no checks at all, `false` means no overlap allowed and `header` means that overlap with the mbr and the gpt header is allowed but nothing else.

Reply:
Ah, once again my imagination fails me - I didn't think of a specially crafted image that would know to jump over a middle part and contain more executable code on the other side.

Yes, some kind of explicit "I know what I'm doing" flag is probably the right way to go. We'll have to decide if both partitions need to "consent" or a unilateral allow-overlap=true from one of them is enough. I think the former is the more robust approach.

Reply:
OK, so I started looking into that "allow some subrange of this partition to overlap/be overwritten by something else". In the end, I decided that the information that a subrange is "don't care" really belongs with the image, and not the partition it is included in - even if the former would be easier to implement. So I added a few patches toward that, including some test cases that I believe should pass, and all but one do - but fixing that one will require adding something other than `pad_file()` to insert the images in the parent image.

I have some idea how such an interface should look, but I'm not sure I'll have something this side of Easter, so I pushed what I have now so people can see where I'm heading.

Reply:
Hm, I gave it a shot. It's a little annoying that I basically had to duplicate all the "loop over extents" logic from pad_file(), but that was a lot simpler than trying to make them share code. Again, all tests (except cramfs, I don't have that) pass after each commit, and there are no XFAILs left now.